sandboxctl.8.in
.\" Copyright 2013 Google Inc.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are
.\" met:
.\"
.\" * Redistributions of source code must retain the above copyright
.\"   notice, this list of conditions and the following disclaimer.
.\" * Redistributions in binary form must reproduce the above copyright
.\"   notice, this list of conditions and the following disclaimer in the
.\"   documentation and/or other materials provided with the distribution.
.\" * Neither the name of Google Inc. nor the names of its contributors
.\"   may be used to endorse or promote products derived from this software
.\"   without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.Dd February 9, 2017
.Dt SANDBOXCTL 8
.Os
.Sh NAME
.Nm sandboxctl
.Nd creates and manages chroot-based sandboxed environments
.Sh SYNOPSIS
.Nm
.Op Fl c Ar config_name
.Op Fl o Ar variable=value
.Op Fl v
config
.Op Ar variable
.Nm
.Op Fl c Ar config_name
.Op Fl o Ar variable=value
.Op Fl v
create
.Nm
.Op Fl c Ar config_name
.Op Fl o Ar variable=value
.Op Fl v
destroy
.Nm
.Op Fl c Ar config_name
.Op Fl o Ar variable=value
.Op Fl v
mount
.Nm
.Op Fl c Ar config_name
.Op Fl o Ar variable=value
.Op Fl v
run
.Ar ...
.Nm
.Op Fl c Ar config_name
.Op Fl o Ar variable=value
.Op Fl v
shell
.Nm
.Op Fl c Ar config_name
.Op Fl o Ar variable=value
.Op Fl v
unmount
.Op Fl f
.Op Fl f
.Sh DESCRIPTION
.Nm
creates and manages chroot-based sandboxed environments in a portable manner
based on a simple configuration file.
.Pp
Because
.Xr chroot 8
is a privileged system command, you must be root to use this tool.
.Pp
The following options apply to all commands:
.Bl -tag -width XXXX
.It Fl c Ar config_name
Specifies the configuration file to use.
The format of the file is described in
.Xr sandboxctl.conf 5 .
.Pp
If
.Ar config_name
includes any directory separator (i.e. one or more slashes) or the
.Sq .conf
suffix, then the argument specifies the path to the configuration file.
.Pp
If
.Ar config_name
is a bare name without any directory components nor the extension, then the
configuration file will be looked up in
.Pa __SANDBOXCTL_ETCDIR__/<config_name>.conf .
The file must exist in this case.
.It Fl o Ar variable=value
Applies an override to the loaded configuration.
.Pp
The
.Ar variable
part of the argument must be any of the recognized configuration variables
described in
.Xr sandboxctl.conf 5 .
The
.Ar value ,
if not empty, specifies the value to set the configuration variable to.
If
.Ar value
is empty, then the configuration variable is unset.
.It Fl v
Enables verbose mode.
.El
.Ss Sandbox types
Sandboxes managed by
.Nm
have a type associated to them.
The sandbox type etermines how the sandbox is constructed and provides all the
operating system-specific knowledge to do so.
There is one or more sandbox types for every supported operating system.
.Pp
For example: on
.Nx ,
there is sandbox type to create an environment that reproduces the host system,
and there also is a sandbox type that recreates a fresh installation of the
system.
Sandbox types are described in
.Xr sandboxctl.conf 5
along with all the settings that affect their behavior.
.Ss Sandbox lifecycle
A sandbox managed by
.Nm
goes through these operations:
.Bl -enum
.It
.Em Configure :
The administrator sets up a configuration file to define the sandbox.
See
.Xr sandboxctl.conf 5
for details.
.It
.Em Create :
Prepares the on-disk layout but does not mount any file systems within it (which
may be necessary depending on the type of the sandbox).
This operation happens once per sandbox.
The sandbox is created with the
.Sq create
command.
.It
.Em Enter or mount :
Performs preparatory work to enter the sandbox, such as mounting any filesystems
within the sandbox or copying dynamic files.
This operation happens every time the sandbox is entered for interaction as
described in the next command.
The sandbox can be manually mounted with the
.Sq mount
command, which is provided for convenience, but you should rarely do this
by hand.
.It
.Em Interact :
The administator can issue commands within the sandbox via the
.Sq run
command or open a shell via the
.Sq shell
command.
.It
.Em Leave or unmount :
Performs the opposite work of the
.Em Enter
step.
In particular, unmounts any file systems that may be left mounted in the
sandbox.
This operation happens every time the interaction step completes.
The sandbox can be manually unmounted with the
.Sq unmount
command, which is provided for convenience, but you should rarely do this
by hand.
.It
.Em Destroy :
The on-disk layout for the sandbox is completely removed.
This operation cannot be executed while there is any active sandbox in the
.Em Interact
stage.
The sandbox is destroyed with the
.Sq destroy
command.
.El
.Pp
Note that sandbox entry/leave operations are reference-counted and can happen
concurrently.
.Nm
ensures that only the first concurrent interaction step executes the entry step
and that only the last concurrent interaction step executes the leave step.
If you get a warning saying that the
.Em sandbox is still in use ,
it means that another instance of
.Nm
is still running for the given sandbox.
.Ss The config command
The config command dumps the loaded configuration to the standard output.
The format of the output is not a script, so it cannot be fed back into
.Nm .
The purpose of this command is to aid in debugging the configuration of the
tool, particularly for those configuration files that use shell logic to
determine the value of the variables they define.
.Pp
A single
.Ar variable
name can be provided as an argument to this command, in which case the value
of this variable is printed verbatim to the standard output.
It is an error if the given variable is not defined.
.Pp
Note that the output of this command varies across operating systems depending
on the sandbox types available.
.Ss The create command
Creates the sandbox, which must not yet exist.
.Ss The destroy command
Destroys the sandbox, which must exist and must not be active.
In other words, no other
.Nm
instance may be running the
.Sq run
or
.Sq shell
commmands.
.Ss The mount command
Mounts sandbox-specific file systems and prepares the sandbox to be usable.
This command is only provided for completeness and convenience; you should not
need to use it by hand: the
.Sq run
and
.Sq shell
commands take care of automatically doing this for you at startup time.
.Ss The run command
Executes a given command inside the sandbox.
The sandbox must exist.
This operation always mounts the sandbox before running the command and unmounts
it after completion.
The shell is forced to
.Pa /bin/sh
and the environment is completely cleaned except for a few key variables.
.Ss The shell command
Executes an interactive shell inside the sandbox.
The sandbox must exist.
This operation always mounts the sandbox before running the shell and unmounts
it after completion.
The shell is forced to
.Pa /bin/sh
and the environment is completely cleaned except for a few key variables.
.Ss The unmount command
Unmounts sandbox-specific file systems.
This command is only provided for completeness and convenience; you should not
need to use it by hand: the
.Sq run
and
.Sq shell
commands take care of automatically doing this for you at exit time.
.Pp
The following options apply to the
.Sq unmount
command:
.Bl -tag -width XXXX
.It Fl f
If given once, causes
.Nm
to disregard any existing locks on the sandbox and unmounts the file systems
within it.
If given twice, causes the unmount operations to also be forced; see the
.Fl f
flag in
.Xr umount 8
for more details about what this involves.
.Pp
You can use this flag to clean up the state of the sandbox if it ever becomes
inconsistent, which is a (rare) possibility because
.Nm
cannot make hard guarantees on the consistency of the sandbox in case of
failures during a regular unmount.
.Pp
.Em Warning :
You should only use this flag if you are certain that no other concurrent
instances of
.Nm
are still running for the given sandbox.
.El
.Sh FILES
.Bl -tag -width XXXX
.It Pa __SANDBOXCTL_ETCDIR__/
Directory containing all system-wide configuration files.
.It Pa __SANDBOXCTL_ETCDIR__/default.conf
Default configuration file to load if the user does not specify any other.
.El
.Sh SEE ALSO
.Xr sandboxctl.conf 5 ,
.Xr chroot 8
.Sh AUTHORS
The
.Nm
utility was developed by
.An Julio Merino
.Aq jmmv@google.com .


sandboxctl.sh
# Copyright 2013 Google Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
# * Redistributions of source code must retain the above copyright
#   notice, this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright
#   notice, this list of conditions and the following disclaimer in the
#   documentation and/or other materials provided with the distribution.
# * Neither the name of Google Inc. nor the names of its contributors
#   may be used to endorse or promote products derived from this software
#   without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

# \file sandboxctl.sh
# Manages sandboxes under various operating systems.

shtk_import cleanup
shtk_import cli
shtk_import config
shtk_import sandbox


# Location of the sandboxctl configuration files.
: ${SANDBOXCTL_ETCDIR:="__SANDBOXCTL_ETCDIR__"}


# Location of the sandboxctl modules.
: ${SANDBOXCTL_MODULESDIR:="__SANDBOXCTL_MODULESDIR__"}


# List of valid configuration variables.
#
# Please remember to update sandboxctl.conf(5) if you change this list.
SANDBOXCTL_CONFIG_VARS="SANDBOX_ROOT SANDBOX_TYPE"


# Sets defaults for configuration variables and hooks that need to exist.
#
# This function should be called before the configuration file has been loaded.
# This means that the user can undefine a required configuration variable, but
# we let him shoot himself in the foot if he so desires.
sandboxctl_set_defaults() {
    # Remember to update sandboxctl.conf(5) if you change any default values.
    shtk_config_set SANDBOX_TYPE "empty"
    sandbox_call_types set_defaults

    post_create_hook() { true; }
    pre_destroy_hook() { true; }
    post_mount_hook() { true; }
    pre_unmount_hook() { true; }
}


# Dumps the loaded configuration.
#
# \param ... The options and arguments to the command.
sandboxctl_config() {
    [ ${#} -le 1 ] || shtk_cli_usage_error "config takes at most one argument"

    if [ ${#} -eq 0 ]; then
        local all_vars="${SANDBOXCTL_CONFIG_VARS} $(sandbox_call_types \
                            config_vars)"
        local sorted_vars="$(for var in ${all_vars}; do \
                                 echo "${var}"; \
                             done | sort | uniq)"
        for var in ${sorted_vars}; do
            if shtk_config_has "${var}"; then
                echo "${var} = $(shtk_config_get "${var}")"
            else
                echo "${var} is undefined"
            fi
        done
    else
        local var="${1}"; shift

        if ! shtk_config_has "${var}"; then
            shtk_cli_error "${var} is not defined"
        fi
        echo "$(shtk_config_get "${var}")"
    fi
}


# Creates the sandbox.
sandboxctl_create() {
    [ ${#} -eq 0 ] || shtk_cli_usage_error "create does not take any arguments"

    local type
    type="$(shtk_config_get SANDBOX_TYPE)" || exit
    local root
    root="$(shtk_config_get SANDBOX_ROOT)" || exit

    [ ! -e "${root}" ] || shtk_cli_error "Sandbox ${root} already exists"
    mkdir "${root}" || shtk_cli_error "Failed to create sandbox root"
    if ! ( sandbox_dispatch "${type}" "${root}" create ); then
        ( sandboxctl_destroy ) || true
        # Cope with the case where the sandbox type was bad: our
        # sandbox_dispatch couldn't do a thing, but sandboxctl_destroy could not
        # run either.  Simply remove the just-created directory.
        rmdir "${root}" 2>/dev/null || true
        return 1
    fi
    if ! ( shtk_config_run_hook post_create_hook ); then
        sandboxctl_destroy || true
        return 1
    fi
}


# Destroys the sandbox.
#
# This does not attempt to unmount the sandbox if still mounted, and will abort
# loudly if it is.
sandboxctl_destroy() {
    [ ${#} -eq 0 ] || shtk_cli_usage_error "destroy does not take any arguments"

    local type
    type="$(shtk_config_get SANDBOX_TYPE)" || exit
    local root
    root="$(shtk_config_get SANDBOX_ROOT)" || exit

    [ -d "${root}" ] || shtk_cli_error "Cannot destroy a non-existent sandbox"

    ! sandbox_has_mounts "${root}" || shtk_cli_error "File systems appear to" \
        "still be mounted; refusing to destroy"

    shtk_config_run_hook pre_destroy_hook
    sandbox_dispatch "${type}" "${root}" destroy

    sandbox_destroy "${root}"
}


# Mounts the sandbox.
#
# The sandbox must have been created first with the 'create' command.  Running
# multiple mount operations from different clients is reasonably safe as we
# record how many clients have called this.
sandboxctl_mount() {
    [ ${#} -eq 0 ] || shtk_cli_usage_error "mount does not take any arguments"

    local type
    type="$(shtk_config_get SANDBOX_TYPE)" || exit
    local root
    root="$(shtk_config_get SANDBOX_ROOT)" || exit

    [ -d "${root}" ] || shtk_cli_error "Cannot mount a non-existent sandbox"

    if sandbox_enter "${root}"; then
        if sandbox_has_mounts "${root}"; then
            sandbox_leave "${root}"
            shtk_cli_error "Sandbox in inconsistent state; mounts found but" \
                           "is not locked"
        fi

        local ret=0
        (
            sandbox_dispatch "${type}" "${root}" mount
            shtk_config_run_hook post_mount_hook
        ) || ret=${?}
        if [ ${ret} -ne 0 ]; then
            sandboxctl_unmount
            exit ${ret}
        fi
    fi
}


# Unmounts the sandbox.
#
# The sandbox must exist.  Running multiple unmount operations from different
# clients is reasonably safe as we have recorded how many clients did so.
sandboxctl_unmount() {
    local force_leave=
    local force_unmount=
    local OPTIND
    while getopts ':f' arg "${@}"; do
        case "${arg}" in
            f)  # Force leave if given once; force unmount if given twice.
                if [ -n "${force_leave}" ]; then
                    force_unmount=-f
                else
                    force_leave=-f
                fi
                ;;

            \?)
                shtk_cli_usage_error "Unknown option -${OPTARG} in unmount"
                ;;
        esac
    done
    shift $((${OPTIND} - 1))
    OPTIND=1  # Should not be necessary due to the 'local' above.

    [ ${#} -eq 0 ] || shtk_cli_usage_error "unmount does not take any arguments"

    local type
    type="$(shtk_config_get SANDBOX_TYPE)" || exit
    local root
    root="$(shtk_config_get SANDBOX_ROOT)" || exit

    [ -d "${root}" ] || shtk_cli_error "Cannot unmount a non-existent sandbox"

    if sandbox_leave ${force_leave} "${root}"; then
        shtk_config_run_hook pre_unmount_hook
        sandbox_dispatch "${type}" "${root}" unmount
        sandbox_unmount_dirs ${force_unmount} "${root}"
    else
        shtk_cli_warning "Sandbox still in use by another process; file" \
            "systems may still be mounted!"
    fi
}


# Executes chroot to enter the sandbox.
#
# \param ... Additional arguments to chroot, if any.
#
# \return The exit status of the executed command.
_sandboxctl_chroot() {
    _SANDBOXCTL_DID_UNMOUNT=false
    unmount_on_signal() {
        if [ "${_SANDBOXCTL_DID_UNMOUNT}" = false ]; then
            sandboxctl_unmount
            _SANDBOXCTL_DID_UNMOUNT=true
        fi
    }
    shtk_cleanup_register unmount_on_signal

    sandboxctl_mount
    local ret=0
    env -i \
        HOME=/tmp \
        PATH="${PATH}" \
        SHELL=/bin/sh \
        TERM="${TERM}" \
        USER="${USER}" \
        chroot "$(shtk_config_get SANDBOX_ROOT)" "${@}" || ret="${?}"
    sandboxctl_unmount
    _SANDBOXCTL_DID_UNMOUNT=true
    return "${ret}"
}


# Runs the given command inside the sandbox.
#
# \param binary Path to the binary to run, relative to the sandbox.
# \param ... Additional arguments to the binary.
#
# \return The exit status of the executed command.
sandboxctl_run() {
    [ ${#} -gt 0 ] || shtk_cli_usage_error "run requires at least one argument"

    _sandboxctl_chroot "${@}"
}


# Runs an interactive shell inside the sandbox.
#
# \return The exit status of the shell.
sandboxctl_shell() {
    [ ${#} -eq 0 ] || shtk_cli_usage_error "shell does not take any arguments"

    PS1="sandbox# " _sandboxctl_chroot
}


# Loads the configuration file specified in the command line.
#
# \param config_name Name of the desired configuration.  It can be either a
#     configuration name (no slashes) or a path.
sandboxctl_config_load() {
    local config_name="${1}"; shift

    local config_file=
    case "${config_name}" in
        */*|*.conf)
            config_file="${config_name}"
            ;;

        *)
            config_file="${SANDBOXCTL_ETCDIR}/${config_name}.conf"
            [ -e "${config_file}" ] \
                || shtk_cli_usage_error "Cannot locate configuration named" \
                "'${config_name}'"
            ;;
    esac
    shtk_config_load "${config_file}"
}


# Entry point to the program.
#
# \param ... Command-line arguments to be processed.
#
# \return An exit code to be returned to the user.
main() {
    local config_name="default"

    sandbox_load_types "${SANDBOXCTL_MODULESDIR}"
    shtk_config_init ${SANDBOXCTL_CONFIG_VARS} $(sandbox_call_types config_vars)

    local OPTIND
    while getopts ':c:o:v' arg "${@}"; do
        case "${arg}" in
            c)  # Name of the configuration to load.
                config_name="${OPTARG}"
                ;;

            o)  # Override for a particular configuration variable.
                shtk_config_override "${OPTARG}"
                ;;

            v)  # Be verbose.
                shtk_cli_set_log_level debug
                ;;

            :)
                shtk_cli_usage_error "Missing argument to option -${OPTARG}"
                ;;

            \?)
                shtk_cli_usage_error "Unknown option -${OPTARG}"
                ;;
        esac
    done
    shift $((${OPTIND} - 1))
    OPTIND=1  # Should not be necessary due to the 'local' above.

    [ ${#} -ge 1 ] || shtk_cli_usage_error "No command specified"

    local exit_code=0

    local command="${1}"; shift
    case "${command}" in
        config|create|destroy|mount|run|shell|unmount)
            sandboxctl_set_defaults
            sandboxctl_config_load "${config_name}"
            "sandboxctl_${command}" "${@}" || exit_code="${?}"
            ;;

        *)
            shtk_cli_usage_error "Unknown command ${command}"
            ;;
    esac

    return "${exit_code}"
}
